### PR TITLE
Refactor the Exportable interface

### PR DESCRIPTION
**Pull request by lanchengx**
Changes proposed in this pull request:
- Adjusted by exporting all the rules that implement this interface to specify the required key for export

### CODE REVIEW

File: shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/rule/DatabaseDiscoveryRule.java [hunk 1]
```diff
@@ -176,9 +177,9 @@ public void updateStatus(final DataSourceStatusChangedEvent event) {
     }
     
     @Override
-    public Map<String, Object> export() {
-        Map<String, Object> result = new HashMap<>(1, 1);
-        result.put(ExportableConstants.PRIMARY_DATA_SOURCE_KEY, exportPrimaryDataSourceMap());
+    public Map<String, Supplier<Object>> getExportMethods() {
+        Map<String, Supplier<Object>> result = new HashMap<>(1);
```
Comment (TeslaCN) [inline shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/rule/DatabaseDiscoveryRule.java:181]: Missing `loadFactor`?

File: shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-distsql/shardingsphere-db-discovery-distsql-handler/src/main/java/org/apache/shardingsphere/dbdiscovery/distsql/handler/query/DatabaseDiscoveryRuleQueryResultSet.java [hunk 2]
```diff
@@ -70,8 +70,9 @@ public void init(final ShardingSphereMetaData metaData, final SQLStatement sqlSt
         discoveryTypes = ruleConfig.map(DatabaseDiscoveryRuleConfiguration::getDiscoveryTypes).orElse(Collections.emptyMap());
         discoveryHeartbeats = ruleConfig.map(DatabaseDiscoveryRuleConfiguration::getDiscoveryHeartbeats).orElse(Collections.emptyMap());
         Optional<ExportableRule> exportableRule = metaData.getRuleMetaData().getRules()
-                .stream().filter(each -> each instanceof ExportableRule).map(each -> (ExportableRule) each).findAny();
-        primaryDataSources = (Map<String, String>) exportableRule.map(optional -> optional.export().get(ExportableConstants.PRIMARY_DATA_SOURCE_KEY)).orElse(Collections.emptyMap());
+                .stream().filter(each -> each instanceof ExportableRule).filter(each -> ((ExportableRule) each).containExportableKey(Collections.singleton(ExportableConstants.PRIMARY_DATA_SOURCE_KEY)))
+                .map(each -> (ExportableRule) each).findAny();
+        primaryDataSources = (Map<String, String>) (exportableRule.map(optional -> optional.export(ExportableConstants.PRIMARY_DATA_SOURCE_KEY)).orElse(Collections.emptyMap()));
```
Comment (TeslaCN) [inline shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-distsql/shardingsphere-db-discovery-distsql-handler/src/main/java/org/apache/shardingsphere/dbdiscovery/distsql/handler/query/DatabaseDiscoveryRuleQueryResultSet.java:75]: Consider using `orElseGet`.

File: shardingsphere-features/shardingsphere-readwrite-splitting/shardingsphere-readwrite-splitting-core/src/main/java/org/apache/shardingsphere/readwritesplitting/rule/ReadwriteSplittingRule.java [hunk 3]
```diff
@@ -118,11 +119,11 @@ public void updateStatus(final DataSourceStatusChangedEvent event) {
     }
     
     @Override
-    public Map<String, Object> export() {
-        Map<String, Object> result = new HashMap<>(1, 1);
-        result.put(ExportableConstants.AUTO_AWARE_DATA_SOURCE_KEY, exportAutoAwareDataSourceMap());
-        result.put(ExportableConstants.AUTO_AWARE_DATA_SOURCE_NAME, exportAutoAwareDataSourceNames());
-        result.put(ExportableConstants.DATA_SOURCE_KEY, exportDataSourceNames());
+    public Map<String, Supplier<Object>> getExportMethods() {
+        Map<String, Supplier<Object>> result = new HashMap<>(2);
```
Comment (TeslaCN) [inline shardingsphere-features/shardingsphere-readwrite-splitting/shardingsphere-readwrite-splitting-core/src/main/java/org/apache/shardingsphere/readwritesplitting/rule/ReadwriteSplittingRule.java:123]: Maybe larger size?

File: shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-distsql/shardingsphere-db-discovery-distsql-handler/src/main/java/org/apache/shardingsphere/dbdiscovery/distsql/handler/query/DatabaseDiscoveryRuleQueryResultSet.java (hunk 2 above)
Comment (TeslaCN) [reply shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-distsql/shardingsphere-db-discovery-distsql-handler/src/main/java/org/apache/shardingsphere/dbdiscovery/distsql/handler/query/DatabaseDiscoveryRuleQueryResultSet.java:76]: Yes, or `orElseGet(Collections::emptyMap)`.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/distsql/constant/ExportableConstants.java [hunk 4]
```diff
@@ -37,4 +37,6 @@
     public static final String REPLICA_DATA_SOURCE_NAMES = "replica_data_source_names";
     
     public static final String DATA_SOURCE_KEY = "data_source_key";
+    
+    public static final String SINGLE_TABLE_TABLES = "single_table_tables";
```
Comment (RaigorJiang) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/distsql/constant/ExportableConstants.java:41]: single_tables

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/rule/identifier/type/ExportableRule.java [hunk 5]
```diff
@@ -19,17 +19,61 @@
...
  */
 public interface ExportableRule extends ShardingSphereRule {
     
     /**
-     * Export data map.
+     * The methods that the rule can supply.
      *
+     * @return export method supplier
+     */
+    Map<String, Supplier<Object>> getExportMethods();
```
Comment (RaigorJiang) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/rule/identifier/type/ExportableRule.java:37]: getExportedMethods  may better?

File: shardingsphere-features/shardingsphere-readwrite-splitting/shardingsphere-readwrite-splitting-core/src/main/java/org/apache/shardingsphere/readwritesplitting/rule/ReadwriteSplittingRule.java [hunk 6]
```diff
@@ -43,6 +43,8 @@
 import java.util.Map;
 import java.util.Optional;
 import java.util.Map.Entry;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
```
Comment (terrymanu) [inline shardingsphere-features/shardingsphere-readwrite-splitting/shardingsphere-readwrite-splitting-core/src/main/java/org/apache/shardingsphere/readwritesplitting/rule/ReadwriteSplittingRule.java:47]: Please remove unused import

**Review by terrymanu — COMMENTED** [review]
src/main/java/org/apache/shardingsphere/readwritesplitting/rule/ReadwriteSplittingRule.java:[47,8] (imports) UnusedImports: Unused import - java.util.stream.Collectors.

**Comment by terrymanu** [comment]
Please resolve conflicting files

**Comment by lanchengx** [comment]
> Please resolve conflicting files

ok, it might have some other modifications

**Review by RaigorJiang — APPROVED** [review]

**Pull request merged**